### ISSUE TITLE
Varints

### ISSUE THREAD

**Issue by bifurcation** [mlswg/mls-protocol#298]
As part of the cTLS work, we looked into the possibility of [using QUIC-style varints as vector lengths](https://github.com/ekr/draft-rescorla-tls-ctls/blob/specialization/draft-rescorla-tls-ctls.md#varints) in the TLS presentation syntax.   With the proper specification text, we could do the same in MLS, as well as using varints for in place of uint values like `epoch`, `sender_index`, and `generation`.

Using varints has a few benefits: 

* It avoids the need to specify vector lengths in the presentation syntax
* It allows any field to go up to 2^64 (possibly not a benefit!)
* It is more efficient at encoding values that are far below the maximum
 
The main drawback is that there's no support for varint vector lengths in existing code for encoding/decoding TLS syntax.

There's also a procedural question of where the varints mechanism should be defined.  Arguably MLS shouldn't be making its own extensions to the TLS syntax.  We've already sort of crossed that bridge, though, since we added optional<T>.

**Comment by bifurcation** [comment]
Better link for [QUIC-style varints](https://tools.ietf.org/html/draft-ietf-quic-transport-22#section-16).  For whatever reason, the cTLS version only does 1/2/3-byte encodings, whereas the QUIC one does 1/2/4/8, which is better for our usage.

**Comment by Bren2010** [comment]
Just repeating what I said during the call, but I think this is mixing layers of abstraction. The MLS spec uses the TLS presentation language to define structs, but we don't actually assume that the wire encoding of messages is TLS. Referencing this line from the architecture doc:

```
In addition, it does not specify a complete wire encoding, but rather a 
set of abstract data structures which can then be mapped onto a variety 
of concrete encodings, such as TLS, CBOR, and JSON.
```

People said that we need to define the wire encoding to support interoperability (if we don't want interop, why are we writing a spec?). But I think in practice, different MLS implementations interoperating will be quite rare for a variety of reasons, mostly due to incompatible delivery and authentication services. The point of writing the spec, then, is to produce a well-reviewed protocol that people can implement confidently.

Long story short, if you wanted varints you would use cTLS as the wire encoding, which doesn't require any changes to the spec.
